### PR TITLE
fix(forms): Call setDisabledState in in reactive directives when appropriate

### LINKED ISSUE #35309: FormControlDirective and FormControlName do not call setDisabledState on ControlValueAcessor instances when the CVA is added and is enabled
# 🐞 bug report

### Affected Package

@angular/forms

### Is this a regression?

Not that I know of.

### Description

When you add a CVA to a form via the FormControlDirective directive or FormControlName directive with reactive forms and the CVA has implemented setDisabledState the the setDisabledState and the FormControl instance is enabled, setDisabledState is not called on the CVA instance, it is however called with the FormControl is disabled.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-issue-repro2-3fagzc

## 🌍  Your Environment

**Angular Version:**
<pre><code>

google3 head

</code></pre>

**Anything else relevant?**

Nope

### PR DESCRIPTION
**Pull request by jermowery**
When a CVA is added a form and that CVA implements setDisabledState the
reactive form directives should call setDisabledState with true when the
FormControl is disabled and false when it is enabled. Currently,
setDisabledState is only called when the FormControl is disabled.

Fixes #35309

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ x] Tests for the changes have been added (for bug fixes / features)
- [ x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
setDisabledState when adding a control to a form is called with true with the FormControl is disabled and not called when the FormControl is enabled.

Issue Number: #35309 

## What is the new behavior?

setDisabledState is called when adding a control to a form with false when the FormControl is enabled and true with the FormControl is disabled.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x ] No

### CODE REVIEW

File: packages/forms/test/value_accessor_integration_spec.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
 {
   describe('value accessors', () => {
 
-    function initTest<T>(component: Type<T>, ...directives: Type<any>[]): ComponentFixture<T> {
+    function initTest<T>(component: Type<T>, ...directives: Array<Type<any>>): ComponentFixture<T> {
```
Comment (AndrewKushnir) [inline packages/forms/test/value_accessor_integration_spec.ts]: We tend to use a short version of type notation wrt arrays, e.g. `string[]` vs `Array<string>`, so we should probably revert this change.

File: packages/forms/test/value_accessor_integration_spec.ts [hunk 2]
```diff
@@ -1056,6 +1057,38 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
...
+            fixture.detectChanges();
+
+            expect(fixture.componentInstance.form.status).toEqual('DISABLED');
+            expect(fixture.componentInstance.form.get('login') !.status).toEqual('DISABLED');
+            expect(fixture.debugElement.query(By.directive(CvaWithDisabledState))
+                       .nativeElement.textContent)
+                .toEqual('DISABLED');
+          });
+
+          it('sets the disabled state when the control is initally enabled', () => {
```
Comment (AndrewKushnir) [inline packages/forms/test/value_accessor_integration_spec.ts]: I'm a bit confused by the test description, should it say `sets the enabled state...` (seems like this is what you are also checking in the test)?

File: packages/forms/test/value_accessor_integration_spec.ts [hunk 3]
```diff
@@ -481,7 +481,8 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
         const assertOptionElementSelectedState = (selectedStates: boolean[]): void => {
           const options = fixture.debugElement.queryAll(By.css('option'));
           if (options.length !== selectedStates.length) {
-            throw 'the selected state values to assert does not match the number of options';
+            throw new Error(
```
Comment (AndrewKushnir) [inline packages/forms/test/value_accessor_integration_spec.ts]: 👍

**Review by AndrewKushnir — CHANGES_REQUESTED** [review]
@jermowery thanks for creating a PR! 👍 I left a couple nits, please have a look.

It looks like this PR also introduces behavior change for Forms, where `setDisabledState` is called in situations where it was not called before. Could you please try to run Global TAP with this change to see if there are targets that might be potentially affected?

Thank you.

**Comment by jermowery** [comment]
> It looks like this PR also introduces behavior change for Forms, where `setDisabledState` is called in situations where it was not called before. Could you please try to run Global TAP with this change to see if there are targets that might be potentially affected?

There are 3 specs failing as a result of this change. 2 of which are in my own codebase. I will put together a CL that can be patched with the fixes for the Angular caretaker doing the sync

**Review by AndrewKushnir — COMMENTED** [review]
Thanks for additional changes @jermowery!

> There are 3 specs failing as a result of this change. 2 of which are in my own codebase. I will put
> together a CL that can be patched with the fixes for the Angular caretaker doing the sync

Thanks for verifying this in g3 👍Given that there are some (minor) breakages, I'd like to ask @kara to have a look at this PR too, to make sure we assess the risks accordingly.

Thank you.

**Comment by jermowery** [comment]
I just took another look at my Global TAP and it looks like there are more changes than anticipated. I will discuss with you all tomorrow

**Comment by sonukapoor** [comment]
Hi @jermowery, I was working on a similar PR, but noticed that your PR already covers what I was going to change as well. Is there anything I can help with for this PR?

**Comment by jermowery** [comment]
Unfortunately, this breaks a significant number of tests inside Google, and I don't have to bandwidth to fix them, so this PR is pretty much abandoned

**Comment by AndrewKushnir** [comment]
FYI, potentially related PR #18575 and related (duplicate?) issue #17806.

**Comment by AndrewKushnir** [comment]
Hi @jermowery, since this is a breaking change and it affects a lot of targets internally in google3, I was thinking if we should consider adding a config option similar to the `warnOnNgModelWithFormControl` one [defined here](https://github.com/angular/angular/blob/d1ea1f4c7f3358b730b0d94e65b00bc28cae279c/packages/forms/src/form_providers.ts#L55-L64). That should keep existing behavior, allow enabling new/corrected one and simplify the rollout of this change (make it incremental). WDYT?

**Comment by jermowery** [comment]
That seems reasonable. Sadly I don't have the bandwidth right now to incorporate that change but feel free to adopt this PR if it is a priority

**Comment by grosch** [comment]
What's the status of this being implemented?

**Comment by mohamed-gara** [comment]
There is already this PR #20310. Maybe it's the same fix.

**Comment by Jakartoine** [comment]
@AndrewKushnir,

Sorry for disturbing you but do you have any update about this point?

To workaround this, I had to recreate a custom radio button component with an override on NG_VALUE_ACCESSOR.

Already 3 years since this bug has been discovered without any fix...

Thank you in advance

**Comment by AndrewKushnir** [comment]
Hi @Jakartoine,

Unfortunately we don't have any updates on this issue at this moment.

The problem with the fix is that it requires a breaking change: it changes the behavior (to be more correct), but this affects existing apps since some of them rely on the current sequence of calls to the `setDisabledState` methods defined on CVAs. Running tests in Google's codebase indicated that the breaking change is quite severe and there is no easy way to auto-correct the code with schematic/migration. We have to come up with a migration strategy, such as introducing a flag (something like `ReactiveFormsModule.withConfig({setDisabledStateBehavior: 'legacy' /* or 'new' */})`) for apps to opt into using the "correct" behavior, wait for 1-2 major versions and only after that flip the flag to make "correct" behavior default. This work requires some planning and additional investigations and we didn't have a chance to prioritize this yet. We'll keep this PR updated once we have more info available.

Thank you.

**Comment by Jakartoine** [comment]
@AndrewKushnir

Alright, it sounds clearer now. Thank you for your response mate.

**Comment by dylhunn** [comment]
I'm going to run a TGP to find out how severe the breakage is. I suspect it's quite bad, but the only way to know is to try!

**Pull request closed without merging**